core/drivers/ns16550.c

```diff
@@ -0,0 +1,37 @@
+/*
+ * NS16550 support
+ * taken from u-boot source (drivers/serial/ns16550.c)
+ * originally from linux source (arch/powerpc/boot/ns16550.c)
+ * modified to use only NS16550_init, NS16550_putc and NS16550_tstc.
+ */
+
+#include <drivers/ns16550.h>
+#include <io.h>
+#define UART_LCRVAL			0x03
+#define UART_MCRVAL			0x03
+#define UART_FCRVAL			0x07
+#define CONFIG_SYS_NS16550_IER		0x00
+#define UART_LSR_THRE			0x20
+#define UART_LCR_BKSE			0x80
+#define serial_out(x, y)	write8(x, (vaddr_t)y)
+#define serial_in(y)		read8((vaddr_t)y)
+
+void NS16550_init(void)
+{
+	/*
+	 * Do nothing, debug ns16550_init share with normal world,
+	 * everything for ns16550_init intialization is done in bootloader.
+	 */
```

Comment:
Then why keep this function? "Driver completeness" might be an answer but in this case we need a proper implementation I think.